### PR TITLE
added unused link cleaner script and readme

### PR DESCRIPTION
**Pull request by radixdev**
# Pull Request/Issue Resolution

**Description of Change:**
Added a script that automatically cleans files of unused shortlinks at the bottom of basically most markdown files in our docs.

### Is this change associated with a Braze feature/product release?
- [ ] Yes (__Insert Feature Release Date Here__)
- [x] No

> If yes, please note the date of the feature release.

---
---

<details>
<summary>PR Checklist</summary>

- [ ] Check that all links work!
- [ ] Ensure you have completed [our CLA](https://www.braze.com/docs/cla/).
- [ ] Tag @Timothy-Kim and @KellieHawks as a reviewer when your work is _done and ready to be reviewed for merge_.
- [ ] Tag others as Reviewers as necessary.
- [ ] If you have modified any links, be sure to add redirects to `config/nginx.conf.erb`.

</details>

<details>
<summary>Writing & Style Quick Tips</summary>

- Use the formatting shown on the [Styling Test Page](https://github.com/Appboy/braze-docs/blob/develop/_docs/_home/styling_test_page.md) to create unique templates and vehicles for your information.
- Use Action verbs!
- Capitalize proper nouns, unique Braze tools and features, custom Braze channels, and all Header titles.
- Redirect when necessary!
- Always speak to the Braze customer - refer to them as “you”, whether they are an Engineer/Developer, a Marketer, or an Admin.
- On Voice: You should aim to be a Friendly Professor. Imagine, if you will, the smartest, but also kindest professor you had. You know, the one who showed you your potential and tested you, but also went slowly in class so you could take great notes and learn some things? Don't be too peppy, but don't be formal. You’re not selling anything - you’re instructing and informing and enabling.

_Consult the [Docs Text Formatting Guide](https://github.com/Appboy/braze-docs/wiki/Special-Formatting) and the [Docs Writing Style Guide & Best Practices](https://github.com/Appboy/braze-docs/wiki/Writing-Style-Guide-&-Best-Practices) for more details._

</details>

<details>
<summary>ATTENTION: For PR Reviewers</summary>

- [ ] Read our [Reviewing a PR page](https://github.com/Appboy/braze-docs/wiki/Reviewing-a-PR) for more on our reviewing suggestions.
- [ ] Preview all changes in the linked Heroku environment (click `View deployment` button below, then "Docs". A `502` error just means you should refresh until you see the Docs Home page.
</details>

---
---

### CODE REVIEW

File: scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+[1]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/initial_sdk_setup/android_sdk_integration
+```
+
+When this link is unused/unreferenced in the actual file, this script will automatically delete it from the file itself. If given a directory, this script will recursively perform this cleaning on all child files of the directory, including inside child directories.
+
+## Usage
+```
+ruby unused_link_cleaner.rb PATH/TO/FILE
+ruby unused_link_cleaner.rb PATH/TO/DIRECTORY
+```
```
Comment (radixdev) [inline scripts/README.md:16]: usage

File: _docs/_developer_guide/platform_integration_guides/android/customer_feedback.md [hunk 2]
```diff
@@ -116,14 +116,6 @@ If you would prefer, you can override these styles to create a look and feel tha
...
-[7]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/customer_feedback/#customer-feedback
-[8]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/customer_feedback/#customer-feedback
-[9]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/customer_feedback/#manual-feedback-collection
-[10]: {{site.baseurl}}/user_guide/data_and_analytics/uninstall_tracking/#implementation
 [13]: http://www.desk.com
 [14]: http://www.zendesk.com
-[16]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/customer_feedback/#third-party-provider-integrations
-[17]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/customer_feedback/#theming-the-braze-feedback-ui
-[18]: {{site.baseurl}}/developer_guide/platform_integration_guides/android/news_feed/#overriding-styles
 [20]: {% image_buster /assets/img_archive/Image31Theming.png %} "Android Feedback"
```
Comment (radixdev) [inline _docs/_developer_guide/platform_integration_guides/android/customer_feedback.md:121]: a taste of the script on an individual file

Comment (radixdev) [reply _docs/_developer_guide/platform_integration_guides/android/customer_feedback.md:121]: If you check the original file, you'll see that `[3]` is actually used, but for example `[6]` was not.

File: scripts/unused_link_cleaner.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    # Search for that ref in the file
+    is_in_file = original_file_lines.grep(/\[#{ref}\]/)
+    if is_in_file.length < 2
+      # This shortlink can be deleted
+      lines_to_remove << current_line[1]
+    end
+  end
+
+  # Since we're deleting the original list, the line number
```
Comment (Bucimis) [inline scripts/unused_link_cleaner.rb:39]: Deletion always goes in order from smallest line to highest right? Might be worth commenting that

**Review by Bucimis — APPROVED** [review]

File: scripts/unused_link_cleaner.rb (hunk 3 above)
Comment (radixdev) [reply scripts/unused_link_cleaner.rb:39]: commented

**Comment by radixdev** [comment]
@KellieHawks @timothy-kim  feel free to merge whenever. I will make a followup pull after this where I use the script to clean ALL docs files in the repo of unused links using the tool.

File: scripts/unused_link_cleaner.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+def is_md_file(full_file_path)
+  return File.extname(full_file_path) == ".md"
+end
+
+def clean_single_file(full_file_path)
+  contents = get_file_contents(full_file_path)
+  lines = contents.lines
+  # find the url shortlinks at the bottom of the file
+  shortlink_lines = get_shortlink_lines(lines)
```
Comment (chrishaines) [inline scripts/unused_link_cleaner.rb:57]: nit: Maybe call this within `#remove_unused_shortlink_lines` so it doesn't have to get passed in as an argument?

**Review by chrishaines — DISMISSED** [review]
Very cool 😎. Just a bunch of nits.

File: scripts/unused_link_cleaner.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  end
+end
+
+if File.directory?(INPUT_FILE)
+  puts "Input directory #{INPUT_FILE}. Will recurse!"
+  puts "Continue? [y/n]"
+  # flush the STDOUT since gradle won't do that for us
+  # Without flushing, the user won't be able to read the prompt
+  STDOUT.flush
+  prompt_response = STDIN.gets.chomp
```
[the remark this answers is not in the document]
Comment (radixdev) [reply scripts/unused_link_cleaner.rb:87]: done

File: scripts/unused_link_cleaner.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  if (!prompt_response.eql? "y")
```
[the remark this answers is not in the document]
Comment (radixdev) [reply scripts/unused_link_cleaner.rb:89]: done

File: scripts/unused_link_cleaner.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+def get_file_contents(file_path)
+  return File.read(file_path)
+end
+
+def grep_with_index(string_lines, query)
+  return string_lines.each_with_index.select{|e,| e =~ query}
+end
+
+def get_shortlink_lines(fileLines)
```
[the remark this answers is not in the document]
Comment (radixdev) [reply scripts/unused_link_cleaner.rb:18]: missed this one, fixed

File: scripts/unused_link_cleaner.rb (hunk 4 above)
Comment (radixdev) [reply scripts/unused_link_cleaner.rb:57]: good call, done

**Comment by radixdev** [comment]
@KellieHawks this is now ready!

**Review by KellieHawks — APPROVED** [review]
lgtm!

**Comment by KellieHawks** [comment]
@cla-bot check

**Pull request merged**